test.py

```diff
@@ -1,131 +1,177 @@
+    ),  # = 1125, Odd `len(T)`, and `len(T) == next_fast_len(len(T), real=True)`.
+    (
+        2 * 7 * 11 * 13,
+        0,
+        lt,
+    ),  # = 2002, Even `len(T)`, and `len(T) < next_fast_len(len(T), real=True)`
+    (
+        7 * 11 * 13,
+        1,
+        lt,
+    ),  # = 1001, Odd `len(T)`, and `len(T) < next_fast_len(len(T), real=True)`
+]
+
+
+def naive_sliding_dot_product(Q, T):
+    m = len(Q)
+    l = T.shape[0] - m + 1
+    out = np.empty(l)
+    for i in range(l):
+        out[i] = np.dot(Q, T[i : i + m])
+    return out
+
+
+@pytest.mark.parametrize("n_T, remainder, comparator", test_inputs)
+def test_remainder(n_T, remainder, comparator):
+    assert n_T % 2 == remainder
+
+
+@pytest.mark.parametrize("n_T, remainder, comparator", test_inputs)
+def test_comparator(n_T, remainder, comparator):
+    shape = next_fast_len(n_T, real=True)
+    assert comparator(n_T, shape)
+
+
+@pytest.mark.parametrize("n_T, remainder, comparator", test_inputs)
+def test_sdp(n_T, remainder, comparator):
+    # test_sdp for cases 1-4
+
+    n_Q_prime = [
+        2,
+        3,
+        5,
+        7,
+        11,
+        13,
+        17,
+        19,
+        23,
+        29,
+        31,
+        37,
+        41,
+        43,
+        47,
+        53,
+        59,
+        61,
+        67,
+        71,
+        73,
+        79,
+        83,
+        89,
+        97,
+    ]
+    n_Q_power2 = [2, 4, 8, 16, 32, 64]
+    n_Q_values = n_Q_prime + n_Q_power2 + [n_T]
+    n_Q_values = sorted(n_Q for n_Q in set(n_Q_values) if n_Q <= n_T)
+
+    modules = utils.import_sdp_mods()
+    for n_Q in n_Q_values:
+        Q = np.random.rand(n_Q)
+        T = np.random.rand(n_T)
+        ref = naive_sliding_dot_product(Q, T)
+        for mod in modules:
+            try:
+                comp = mod.sliding_dot_product(Q, T)
+                npt.assert_allclose(comp, ref)
+            except Exception as e:  # pragma: no cover
+                msg = f"Error in {mod.__name__}, with n_Q={n_Q} and n_T={n_T}"
+                print(msg)
+                raise e
+
+    return
+
+
+def test_sdp_power2():
+    # test for case 5. len(T) is power of 2
+    pmin = 3
+    pmax = 13
+
+    modules = utils.import_sdp_mods()
+    for mod in modules:
+        try:
+            for q in range(pmin, pmax + 1):
+                n_Q = 2**q
+                for p in range(q, pmax + 1):
+                    n_T = 2**p
+                    Q = np.random.rand(n_Q)
+                    T = np.random.rand(n_T)
+
+                    ref = naive_sliding_dot_product(Q, T)
+                    comp = mod.sliding_dot_product(Q, T)
+                    npt.assert_allclose(comp, ref)
+
+        except Exception as e:  # pragma: no cover
+            msg = f"Error in {mod.__name__}, with q={q} and p={p}"
+            print(msg)
+            raise e
+
+    return
+
+
+def test_setup():
+    Q = np.random.rand(3)
+    T = np.random.rand(10)
+
+    modules = utils.import_sdp_mods()
```

Comment:
@NimaSarajpoor 
This test does not make sense because `utils.import_sdp_mods()` checks modules and only import them if they have the function `setup`.  So, the actual test we are doing here is only for the returned value.  We want to check ALL modules in the `spd/` sand make sure they ALL have the function `setup`. 

**Next step:** 
Revise the test function by checking all modules in `sdp` and make sure it has both functions `sliding_dot_product` and `setup`. (Its downside is that we cannot easily create and use any helper module in sdp for `challenger_sdp`)